src/optimizers.py

```diff
@@ -45,85 +45,91 @@ def optimize(self, net_inp, target):
         net_outputs = self.__nn.forward(inp=net_inp)
         err = self.loss.func(predicted=net_outputs, target=target)
 
-        # 1) per step
-        d_err = self.loss.deriv(predicted=net_outputs, target=target)
-        for i in reversed(range(len(self.__nn.layers))):
+        output_layer = self.__nn.layers[-1]
+        output_act = output_layer.act
+
+        # dErr_dOut: gradient of the error wrt the net's outputs
+        # d_out: gradient of the net's outputs wrt the output units' weighted sums
+        # dNet_dOut: gradient of the output units' weighted sum wrt the prev layer's outputs
+        dErr_dOut = self.loss.deriv(predicted=net_outputs, target=target)
+        d_out = [output_act.deriv(u.net) for u in output_layer.units]
+        dNet_dOut = [u.w[j] for u in output_layer.units for j in range(len(u.w))]
+
+        # delta_next is the delta of the output layer
+        # 'next' is because it will be continuously overwritten with the delta of the next layer
+        # it's normal mult because the vecs have the same dimension and they are 'numpy ndarray'
+        delta_next = dErr_dOut * d_out
+
+        # will contain all the delta_w to update the weights
+        delta_w = [None] * len(self.__nn.layers)
+
+        # scan all layers from the penultimate to the first
+        for i in reversed(range(len(self.__nn.layers) - 1)):
             curr_layer = self.__nn.layers[i]
+            next_layer = self.__nn.layers[i + 1]
+
+            # gradient of the error wrt the outputs of the CURRENT layer
+            dErr_dOut_new = np.zeros([len(curr_layer.units)])
+            for j in range(len(curr_layer.units)):
+                for l in range(len(next_layer.units)):
+                    # dNet_dOut[offset * l + j]: weight (deriv of net wrt out) on the connection j --> l
+                    dErr_dOut_new[j] += dErr_dOut[l] * d_out[l] * next_layer.units[l].w[j]
```

Comment:
This part was not so clear. 
if we are in the output layer we have: dE/dw = delta * d_net/dw    (where delta = dE/d_out * d_out/d_net --> basically the delta_next you wrote in the code). 
Until here everything ok!

d_net/dw = out of the previous layers taken by the neuron in which we are backprop in, until we reach the input vector and here the d_net/dw is simply the input vector x.
Now we have to calculate delta. 
For output layer it's ok is just the deriv of the err for that neuron * the deriv of the act function 
Now when we move to units in the hidden layer here the things change wrt dE/d_out --> because now:
dE/d_out = dE/d_net * d_net/d_out 
Now dE/d_net = dE/out * d_out/d_net --> calculated in the previous layer ( basically the delta_next that will be overwritten after each iteration ...right???)

The last piece for calculating dE/d_out is d_net /d_out which is basically the weight we want to update.

- At line 72 you initialize dE/d_out to zero ( but I dont get why you put "new") 

- Then I'm completely lost at line 76. (can you explain what you did there?)

Reply:
> Now dE/d_net = dE/out * d_out/d_net --> calculated in the previous layer ( basically the delta_next that will be overwritten after each iteration ...right???)

Yes, it's overwritten at each iteration.
<br>

> At line 72 you initialize dE/d_out to zero ( but I dont get why you put "new")

Because `dErr_dOut` refers to the **previous layer's** derivative of the error.
Since it's needed to compute the derivative of the error of the **current layer**, I need 2 distinct variables.
Then the initialization to 0 is because I allocate the memory for the whole vector, otherwise I wouldn't be able to do assignments in specific positions (like `dErr_dOut_new[j] = something`)
<br>

> Then I'm completely lost at line 76. (can you explain what you did there?)

So the line of code is: `dErr_dOut_new[j] += dErr_dOut[l] * d_out[l] * next_layer.units[l].w[j]`
* Position `j` refers to the unit in the **current layer**.
* Position `l` refers to the unit in the **previous layer**.

The deriv of the error wrt the output of the **current layer** is equal to the **sum** of:
- the deriv of the error wrt the output of the **following layer** times
- the deriv of the output wrt the weighted sum of the current layer times
- the weight `ij`: the one on the connection between the current unit `j` in the current layer and the current unit `l` in the following layer

The line in question is actually equivalent to:
```
offset = len(curr_layer.units)
dErr_dOut_new[j] += dErr_dOut[l] * d_out[l] * dNet_dOut[offset * l + j]
```